### PR TITLE
Fix WPS115 for enum attributes #3238

### PR DESCRIPTION
**Pull request by Tapeline**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Refs #3238

### CODE REVIEW

File: tests/test_visitors/test_ast/test_naming/test_class_attributes.py [hunk 1]
```diff
@@ -31,6 +31,30 @@ def action_method(self, request, second):
...
 
+static_attribute_in_enum = """
+import enum
+
+
+class Test(enum.Enum):
+    {0} = enum.auto()
+"""
+
+static_attribute_in_django_enum_like = """
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_naming/test_class_attributes.py:42]: I would go the other way. I would provide a template like:

```python
template = """
class MyClass({0}):
    {1}
"""
```

and then format it with bases like:
- `choices.TextChoices`
- `Enum`
- `enum.Enum`
- etc

And bodies like:
- `UPPER: str = 1`
- `UPPER = ...`
- `UPPER_NAME = UPPER_ALIAS = auto()`
- `lower_also_fine: int = 1`
- `lower_also_fine = 1`
- etc

and check that there are no violations. Then, I would check that non-enum classes cannot have `UPPER_NAMES` with changing the base class:
- `not_enum.Base`
- `metaclass=CustomMeta`
- etc

And names:
- `UPPER_NAME: int = 1`
- etc

This cases must raise a violation.

File: wemake_python_styleguide/logic/naming/enums.py [hunk 2]
```diff
@@ -12,8 +13,34 @@
...
-    string_bases = {node_to_string(base) for base in defn.bases}
-    return any(enum_base in string_bases for enum_base in _ENUM_NAMES)
+    return _has_one_of_base_classes(defn, _ENUM_NAMES)
+
+
+def has_enum_like_base(defn: ast.ClassDef) -> bool:
+    """
+    Tells if some class has `Enum` or semantically similar class as its base.
+
+    Unlike has_enum_base it also includes support for Django Choices.
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/naming/enums.py:44]:
```suggestion
    Unlike ``has_enum_base`` it also includes support for Django Choices.
    https://docs.djangoproject.com/en/5.1/ref/models/fields/#choices
```

File: wemake_python_styleguide/violations/naming.py [hunk 3]
```diff
@@ -488,6 +488,10 @@ class UpperCaseAttributeViolation(ASTViolation):
     """
     Require ``snake_case`` for naming class attributes.
 
+    Exceptions:
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py:491]:
```suggestion
```

Just explain this in the body, no need for an extra heading.

File: wemake_python_styleguide/violations/naming.py [hunk 4]
```diff
@@ -503,6 +507,13 @@ class UpperCaseAttributeViolation(ASTViolation):
         class A:
             my_attribute = 42
 
+        # Correct:
+        import enum
+
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/naming.py:512]:
```suggestion
```

File: wemake_python_styleguide/visitors/ast/naming/validation.py [hunk 5]
```diff
@@ -22,6 +22,9 @@
     logical,
     name_nodes,
 )
+from wemake_python_styleguide.logic.naming.enums import (
+    has_enum_like_base,
+)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming/validation.py:27]:
```suggestion
```

File: wemake_python_styleguide/visitors/ast/naming/validation.py [hunk 6]
```diff
@@ -22,6 +22,9 @@
     logical,
     name_nodes,
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming/validation.py:23]:
```suggestion
    name_nodes,
    enums,
```

File: wemake_python_styleguide/visitors/ast/naming/validation.py [hunk 7]
```diff
@@ -227,14 +230,20 @@ def check_attribute_names(self, node: ast.ClassDef) -> None:
...
-                    self._ensure_case(assign, attr_name)
+                    self._ensure_case(assign, attr_name, is_enum_like)
 
-    def _ensure_case(self, node: AnyAssign, name: str) -> None:
-        if logical.is_upper_case_name(name):
+    def _ensure_case(
+        self,
+        node: AnyAssign,
+        name: str,
+        is_enum_like: bool,  # noqa: FBT001
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming/validation.py:244]:
```suggestion
        name: str,
        *,
        is_enum_like: bool,  # noqa: FBT001
```

File: wemake_python_styleguide/visitors/ast/naming/validation.py [hunk 8]
```diff
@@ -227,14 +230,20 @@ def check_attribute_names(self, node: ast.ClassDef) -> None:
             node,
             include_annotated=True,
         )
+        is_enum_like = has_enum_like_base(node)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming/validation.py:233]:
```suggestion
        is_enum_like = enums.has_enum_like_base(node)
```

File: wemake_python_styleguide/visitors/ast/naming/validation.py [hunk 9]
```diff
@@ -227,14 +230,20 @@ def check_attribute_names(self, node: ast.ClassDef) -> None:
...
 
         for assign in class_attributes:
             for target in get_assign_targets(assign):
                 for attr_name in name_nodes.get_variables_from_node(target):
-                    self._ensure_case(assign, attr_name)
+                    self._ensure_case(assign, attr_name, is_enum_like)
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/naming/validation.py:238]:
```suggestion
                    self._ensure_case(assign, attr_name, is_enum_like=is_enum_like,)
```

**Review by sobolevn — COMMENTED** [review]
Thank you!

File: tests/test_visitors/test_ast/test_naming/test_class_attributes.py (hunk 1 above)
Comment (Tapeline) [reply tests/test_visitors/test_ast/test_naming/test_class_attributes.py:42]: > Then, I would check that non-enum classes cannot have UPPER_NAMES with changing the base class:

Isn't that covered already by ``test_upper_case_class_attributes``?

Comment (Tapeline) [reply tests/test_visitors/test_ast/test_naming/test_class_attributes.py:42]: Also, maybe reuse `logic.naming.enums._ENUM_LIKE_NAMES` in test cases rather that manually writing this?

```py
@pytest.mark.parametrize(
    'base_class',
    [
        'Enum',
        'enum.Enum',
        'models.IntegerChoices',
        'IntegerChoices',
        'models.TextChoices',
        'TextChoices',
        'models.Choices',
        'Choices',
    ],
)
```
Transform to 
```py
@pytest.mark.parametrize(
    'base_class',
    _ENUM_LIKE_NAMES,
)
```

File: tests/test_visitors/test_ast/test_naming/test_class_attributes.py [hunk 10]
```diff
@@ -31,29 +32,15 @@ def action_method(self, request, second):
...
-class TestChoice2(models.IntegerChoices):
-    {0} = 1
-
-
-class TestChoice3(models.Choices):
-    {0} = True
-"""
+upper_case_attributes = [
+    'UPPER: str = 1',
+    'UPPER_CASE = ...',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_naming/test_class_attributes.py:42]:
```suggestion
    'UPPER_CASE = ...',
    'lower = UPPER = auto()',
    'lower, UPPER = field',
```

**Review by sobolevn — COMMENTED** [review]
Awesome! Thank you!

**Review by sobolevn — APPROVED** [review]
Thanks!

**Pull request merged**